src/model.py

```diff
@@ -304,46 +315,202 @@ def validate_device_config(self):
-
-    async def await_shutdown(self):
-        """
-        Primary coroutine running on the engine event loop. This coroutine is responsible for
-        keeping the engine alive until a shutdown is requested.
-        """
-        # first await the shutdown signal
-        while self._shutdown_event.is_set() is False:
-            await asyncio.sleep(5)
-
-        # Wait for the ongoing_requests
-        while self.ongoing_request_count > 0:
-            self.logger.log_info(
-                "[vllm] Awaiting remaining {} requests".format(
-                    self.ongoing_request_count
+    def _setup_lora(self):
+        self.enable_lora = False
+
+        # Check if `enable_lora` field is in the `model.json`,
+        # and if it is, read its contents, which can be string or bool.
+        if (
+            "enable_lora" in self.vllm_engine_config.keys()
+            and str(self.vllm_engine_config["enable_lora"]).lower() == "true"
+        ):
+            # create Triton LoRA weights repository
+            multi_lora_args_filepath = os.path.join(
+                pb_utils.get_model_dir(), _MULTI_LORA_ARGS_FILENAME
+            )
+            try:
+                with open(multi_lora_args_filepath) as lora_file:
+                    lora_repository: Dict[str, str] = json.load(lora_file)
+                self.lora_repository = lora_repository
+                self.supported_loras: List[str] = list(self.lora_repository.keys())
+                self.supported_loras_len = len(self.supported_loras)
+                self.enable_lora = True
+            except FileNotFoundError:
+                raise FileNotFoundError(
+                    f"Triton backend cannot find {multi_lora_args_filepath}."
+                )
+
+    def _setup_metrics(self):
+        self._vllm_metrics = None
+        # TODO: Do not read metrics directly from the vLLM engine, read from prometheus
+        #       client to allow the use of ZMQ process when metrics are enabled. See
+        #       https://github.com/vllm-project/vllm/blob/v0.6.3.post1/vllm/entrypoints/openai/api_server.py#L222-L245
+        if self._enable_metrics:
+            try:
+                labels = {
+                    "model": self.args["model_name"],
+                    "version": self.args["model_version"],
+                }
+                # Add vLLM custom metrics
+                engine_config = self._llm_engine.engine.model_config
+                self._vllm_metrics = VllmStatLogger(
+                    labels, engine_config.max_model_len, self.logger
+                )
+                self._llm_engine.add_logger("triton", self._vllm_metrics)
+            except pb_utils.TritonModelException as e:
+                if "metrics not supported" in str(e):
+                    # Metrics are disabled at the server
+                    self.logger.log_info("[vllm] Metrics not supported")
+                else:
+                    raise e
```

Comment:
Can we do some re-factoring in the logic, please.

1) Let's call `_setup_metrics` prior to `_init_engine`
2) Let's initialize `self._enable_metrics` in `_setup_metrics`
3) Do something like :

```python
    def _setup_metrics(self):
        self._vllm_metrics = None
        
        self._enable_metrics = (
            self._get_bool_config_param("REPORT_CUSTOM_METRICS")
            and not self._aync_engine_args.disable_log_stats
        )
        if !self.enable_metrics:
            return
        # TODO: Do not read metrics directly from the vLLM engine, read from prometheus
        #       client to allow the use of ZMQ process when metrics are enabled. See
        #       https://github.com/vllm-project/vllm/blob/v0.6.3.post1/vllm/entrypoints/openai/api_server.py#L222-L245
        try:
                labels = {
                    "model": self.args["model_name"],
                    "version": self.args["model_version"],
                }
                # Add vLLM custom metrics
                engine_config = self._llm_engine.engine.model_config
                self._vllm_metrics = VllmStatLogger(
                    labels, engine_config.max_model_len, self.logger
                )
                self._llm_engine.add_logger("triton", self._vllm_metrics)
          except pb_utils.TritonModelException as e:
                if "metrics not supported" in str(e):
                    # Metrics are disabled at the server
                    self.logger.log_info("[vllm] Metrics not supported")
                else:
                    raise e
```

Reply:
this way all metric re-lated staff will be under the same related function and we'll somewhat follow single responsibility idea

Reply:
or do we need an engine set up before we can set up metrics?

Reply:
discussed offline, I'll take care of re-factor in a follow up PR